src/TParsec/Result.idr

```diff
@@ -13,18 +13,28 @@ data Result : Type -> Type -> Type where
   Value : a -> Result e a
 
 Functor (Result e) where
-  map _ (HardFail e) = HardFail e
   map _ (SoftFail e) = SoftFail e
+  map _ (HardFail e) = HardFail e
   map f (Value a)    = Value (f a)
 
 Applicative (Result e) where
   pure = Value
   (HardFail e) <*> _ = HardFail e
-  (SoftFail e) <*> _ = SoftFail e
   _ <*> (HardFail e) = HardFail e
+  (SoftFail e) <*> _ = SoftFail e
   _ <*> (SoftFail e) = SoftFail e
```

Comment:
I think this re-ordering will result in slower runtimes: now if we get a softfail on the
first computation, we still evaluate the right one instead of returning early. This doesn't
seem right.

Reply:
It also probably breaks the monad-applicative compatibility rules: in the Monad instance we
never run the function if the argument returns a `SoftFail` (and that makes sense: we would not
know what to run it on).